controllers/rollout/canary.go

```diff
@@ -0,0 +1,269 @@
+/*
+Copyright 2022 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rollout
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kruiseappsv1alpha1 "github.com/openkruise/kruise-api/apps/v1alpha1"
+	appsv1alpha1 "github.com/openkruise/rollouts/api/v1alpha1"
+	"github.com/openkruise/rollouts/controllers/rollout/batchrelease"
+	"github.com/openkruise/rollouts/pkg/util"
+	apps "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *rolloutContext) runCanary() error {
+	canaryStatus := r.newStatus.CanaryStatus
+	// In case of continuous publishing(v1 -> v2 -> v3), then restart publishing
+	if canaryStatus.CanaryRevision == "" {
+		canaryStatus.CurrentStepState = appsv1alpha1.CanaryStepStateUpgrade
+		canaryStatus.CanaryRevision = r.workload.CanaryRevision
+	}
+	if canaryStatus.CurrentStepIndex == 0 {
+		canaryStatus.CurrentStepIndex = 1
```

Comment:
why we hack the index here ?

Reply:
Because of the step, we are from 1-n, not 0-n. when canaryStatus.CurrentStepIndex indicates we should start the '1' step.